### PR TITLE
Add exons stats to gene overview

### LINKED ISSUE #366: Add exon coverage to the gene overiew
To display coverage over exons as well exons on the same gene overview page

### PR DESCRIPTION
**Pull request by northwestwitch**
### Added/Changed/Fixed
- Closes #366
- Adds MANE badges on gene overview
- Adds a `Create PDF` button on MANE overview and gene overview pages

image

### How to test
- [x] Deploy on stage and make sure gene overview contains both transcripts and exons intervals with stats

## Review
- [x] Tests executed by CR, Jakob37 
- [x] "Merge and deploy" approved by Jakob37 

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by northwestwitch** [comment]
This PR happens to work for many genes, but there are some (for instance gene MTHFR) that cause the page to load for a long time and eventually to crash with the following error:

```
  File "/Users/chiararasi/Documents/work/GITs/chanjo2/src/chanjo2/meta/handle_d4.py", line 181, in get_gene_overview_stats
    transcripts_completeness = get_d4tools_intervals_completeness(
  File "/Users/chiararasi/Documents/work/GITs/chanjo2/src/chanjo2/meta/handle_completeness_stats.py", line 15, in get_d4tools_intervals_completeness
    d4tools_stats_perc_cov: str = subprocess.check_output(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/chanjo2/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/opt/homebrew/Caskroom/miniconda/base/envs/chanjo2/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['d4tools', 'stat', '-s', 'perc_cov=10,15,20,50,100', '--region', '/var/folders/9x/2v5jk6391t3dbdsr7kpnl4t80000gp/T/tmp705phl2y', '/Users/chiararasi/Documents/work/d4_data/hg002.d4']' died with <Signals.SIGSEGV: 11>.
```

I suspect this might be another problem in d4tools. I'll investigate using d4tools directly on the intervals of this gene

**Comment by northwestwitch** [comment]
> This PR happens to work for many genes, but there are some (for instance gene MTHFR) that cause the page to load for a long time and eventually to crash with the following error:
> 
> ```
>   File "/Users/chiararasi/Documents/work/GITs/chanjo2/src/chanjo2/meta/handle_d4.py", line 181, in get_gene_overview_stats
>     transcripts_completeness = get_d4tools_intervals_completeness(
>   File "/Users/chiararasi/Documents/work/GITs/chanjo2/src/chanjo2/meta/handle_completeness_stats.py", line 15, in get_d4tools_intervals_completeness
>     d4tools_stats_perc_cov: str = subprocess.check_output(
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/chanjo2/lib/python3.8/subprocess.py", line 415, in check_output
>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/chanjo2/lib/python3.8/subprocess.py", line 516, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['d4tools', 'stat', '-s', 'perc_cov=10,15,20,50,100', '--region', '/var/folders/9x/2v5jk6391t3dbdsr7kpnl4t80000gp/T/tmp705phl2y', '/Users/chiararasi/Documents/work/d4_data/hg002.d4']' died with <Signals.SIGSEGV: 11>.
> ```
> 
> I suspect this might be another problem in d4tools. I'll investigate using d4tools directly on the intervals of this gene

Never mind, I was apparently still using the old d4tools, and after I upgraded:

image

Got the gene page working:

image

I'll mark this PR as ready for review then! 😀

**Comment by Jakob37** [comment]
> Never mind, I was apparently still using the old d4tools, and after I upgraded:

OK relief 😅 

> I'll mark this PR as ready for review then! 😀

Great, I'll check!

**Comment by northwestwitch** [comment]
> OK relief 😅

Me too, sorry! 

image

Thanks!

**Comment by Jakob37** [comment]
The MANE select marking works, and I get coverage for the exons. Nice! Did not take too long either, maybe ~10 seconds.

gene_overview_1

gene_overview_2

Something I think the users would want here is a way to easily see what exons belong together with what transcript. I suspect they will want to go to a specific transcript with X% coverage at 10x (in our case - the MANE transcript), and then see which of its exons are dropping in coverage.

I suspect they also would be interested in coordinates / size as well in relation to the transcript.

You could imagine for instance a drop down at each transcript "show exons", folding out a compact table with the relevant exons with some more info.

Or if you have other ideas on how a user easily could navigate -> transcript -> transcript's exons.

What do you think? 🤔

**Comment by northwestwitch** [comment]
> Something I think the users would want here is a way to easily see what exons belong together with what transcript. I suspect they will want to go to a specific transcript with X% coverage at 10x (in our case - the MANE transcript), and then see which of its exons are dropping in coverage.
> 
> (I suspect they also would be interested in coordinates / size as well of the exon. But that is more a speculation, could be left until we know it is something they want).
> 
> You could imagine for instance a drop down at each transcript "show exons", folding out a compact table with the relevant exons with some more info.
> 
> Or if you have other ideas on how a user easily could navigate -> transcript -> transcript's exons.
> 
> What do you think? 🤔

Very good suggestions! I'm looking at what is currently saved in the database and there is no connection between transcripts and exons:

image

So bad news is that to link them we need to change the structure of the exons' table by including another field (ensembl_transcript_id). Too bad I didn't think about it before. Anyway, the infiles used to populate the database do contain this info so it's very doable! (in another PR perhaps)

image

I was already think about coordinates and sizes and that should be quite straightforward. I'll add those to this PR!

**Comment by Jakob37** [comment]
> So bad news is that to link them we need to change the structure of the exons' table by including another field (ensembl_transcript_id). Too bad I didn't think about it before. Anyway, the infiles used to populate the database do contain this info so it's very doable! (in another PR perhaps)

OK I see! Yes makes sense. It is hard to see into the future!

> I was already thinking about coordinates and sizes and that should be quite straightforward. I'll add those to this PR!

OK, nice :)

**Comment by northwestwitch** [comment]
Ok, I have something that seems to works locally and creates a nice recap of all transcripts with their exons (in an expandable div), together with coordinates and sizes:


In order to test it unfortunately you'd need to modify the exons table (which now contains 2 more fields: exon rank and transcript ID) - perhaps it's easier to drop and re-create the database?

Anyway: the page looks like this:

image

**Comment by Jakob37** [comment]
> Ok, I have something that seems to works locally and creates a nice recap of all transcripts with their exons (in an expandable div), together with coordinates and sizes:
> 
> In order to test it unfortunately you'd need to modify the exons table (which now contains 2 more fields: exon rank and transcript ID) - perhaps it's easier to drop and re-create the database?
> 
> Anyway: the page looks like this:
> image

The screenshot looks very nice 👌 

I remember I needed to fiddle a bit with the db last time due to exon download timing out. Will see if I can re-figure it out :)

I think I'll be able to look into this first thing tomorrow!

**Comment by northwestwitch** [comment]
> I remember I needed to fiddle a bit with the db last time due to exon download timing out. Will see if I can re-figure it out :)
> 
> I think I'll be able to look into this first thing tomorrow!

If it times out try downloading the file and providing the path to it instead

**Comment by Jakob37** [comment]
I got it running after some fiddling. 

chanjo2_db

Looks nice and useful 😄 Numbers looks correct as well.

exons

File: src/chanjo2/templates/gene-overview.html [hunk 1]
```diff
@@ -1,41 +1,111 @@
...
+{% endblock %}
+
+
+{% macro one_interval_table(interval_id, interval_features) %}
+<tr>
+	<td class="row">
+		<div class="panel-default">
+			<div class="panel-heading {% if interval_features.interval_type == 'exon' %} exon_blue {% endif %}" >
+				{{ interval_features.interval_type|capitalize }} <strong>{{ interval_id }}</strong>
+				<span> - Coordinates:{{interval_features.coordinates}}</span>
```
Comment (Jakob37) [inline src/chanjo2/templates/gene-overview.html:33]: What about:

```suggestion
				<span> - Coordinates: {{interval_features.coordinates}}</span>
```

File: src/chanjo2/templates/gene-overview.html [hunk 2]
```diff
@@ -1,41 +1,111 @@
...
+				<span> - size:{{interval_features.length}}</span>
```
Comment (Jakob37) [inline src/chanjo2/templates/gene-overview.html:34]:
```suggestion
				<span> - size: {{interval_features.length}}</span>
```

**Review by Jakob37 — APPROVED** [review]
Looks great 🎉

**Comment by northwestwitch** [comment]
> Looks great 🎉

COOL, let's merge as soon as the tests pass!

**Pull request merged**